File: examples/llm-streaming-cutoff/python/mock_server.py
"""OpenAI-compatible mock returning an SSE stream that completes properly."""

import http.server
import json
import threading


_TOKENS = ["The", " final", " answer", " is", " 42", "."]


class _Handler(http.server.BaseHTTPRequestHandler):
    def do_POST(self):
        if self.path != "/v1/chat/completions":
            self.send_error(404)
            return
        length = int(self.headers.get("Content-Length", 0))
        self.rfile.read(length)

        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-cache")
        self.send_header("Connection", "close")
        self.end_headers()

        for token in _TOKENS:
            event = {
                "id": "chatcmpl-test",
                "object": "chat.completion.chunk",
                "model": "gpt-4o-mini",
                "choices": [
                    {
                        "index": 0,
                        "delta": {"content": token},
                        "finish_reason": None,
                    }
                ],
            }
            self.wfile.write(f"data: {json.dumps(event)}\n\n".encode())
            self.wfile.flush()

        final = {
            "id": "chatcmpl-test",
            "object": "chat.completion.chunk",
            "model": "gpt-4o-mini",
            "choices": [{"index": 0, "delta": {}, "finish_reason": "stop"}],
        }
        self.wfile.write(f"data: {json.dumps(final)}\n\n".encode())
        self.wfile.write(b"data: [DONE]\n\n")
        self.wfile.flush()

    def log_message(self, *_):
        pass


def serve_in_background():
    server = http.server.HTTPServer(("127.0.0.1", 0), _Handler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server.server_address[1], server.shutdown

File: examples/llm-streaming-cutoff/nodejs/mock-server.js
// OpenAI-compatible mock returning an SSE stream that completes
// properly. Under the streaming-cutoff fault, the proxy truncates
// the stream after N data events without sending [DONE].
import http from "node:http";

const TOKENS = ["The", " final", " answer", " is", " 42", "."];

export function serveInBackground() {
  const server = http.createServer((req, res) => {
    if (req.url !== "/v1/chat/completions") {
      res.writeHead(404).end();
      return;
    }
    req.on("data", () => {});
    req.on("end", () => {
      res.writeHead(200, {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
        "Connection": "close",
      });
      for (const token of TOKENS) {
        const event = {
          id: "chatcmpl-test",
          object: "chat.completion.chunk",
          model: "gpt-4o-mini",
          choices: [{ index: 0, delta: { content: token }, finish_reason: null }],
        };
        res.write(`data: ${JSON.stringify(event)}\n\n`);
      }
      const final = {
        id: "chatcmpl-test",
        object: "chat.completion.chunk",
        model: "gpt-4o-mini",
        choices: [{ index: 0, delta: {}, finish_reason: "stop" }],
      };
      res.write(`data: ${JSON.stringify(final)}\n\n`);
      res.end("data: [DONE]\n\n");
    });
  });
  return new Promise((resolve) => {
    server.listen(0, "127.0.0.1", () => {
      resolve({
        port: server.address().port,
        shutdown: () => new Promise((r) => server.close(r)),
      });
    });
  });
}
